Modules/CalcPerform.lua

```diff
@@ -14,9 +14,308 @@ local m_ceil = math.ceil
+				ConPrintf(skill.activeEffect.grantedEffect.name)
+				return skill, GlobalCache.cachedData["CACHE"][uuid].Speed
+			elseif not reqManaCost then
+				return skill, GlobalCache.cachedData["CACHE"][uuid].Speed
+			end
+		end
+	end
+	return source, triggerRate
+end
+
+-- Calculate Trigger Rate Cap accounting for ICDR
+local function getTriggerActionTriggerRate(baseActionCooldown, env, breakdown, focus)
+	local icdr = 1
+	if focus then
+		icdr = calcLib.mod(env.player.mainSkill.skillModList, env.player.mainSkill.skillCfg, "FocusCooldownRecovery")
+		env.player.mainSkill.skillData.focussed = true
+	else
+		icdr = calcLib.mod(env.player.mainSkill.skillModList, env.player.mainSkill.skillCfg, "CooldownRecovery")
+	end
+
+	local modActionCooldown = baseActionCooldown / (icdr)
+	local rateCapAdjusted = m_ceil(modActionCooldown * data.misc.ServerTickRate) / data.misc.ServerTickRate
+	local extraICDRNeeded = m_ceil((modActionCooldown - rateCapAdjusted + data.misc.ServerTickTime) * icdr * 1000)
+	if breakdown then
+		breakdown.ActionTriggerRate = {
+			s_format("%.2f ^8(base cooldown of triggered skill)", baseActionCooldown),
+			s_format("/ %.2f ^8(increased/reduced cooldown recovery)", icdr),
+			s_format("= %.4f ^8(final cooldown of trigger)", modActionCooldown),
+			s_format(""),
+			s_format("%.3f ^8(adjusted for server tick rate)", rateCapAdjusted),
+			s_format("^8(extra ICDR of %d%% would reach next breakpoint)", extraICDRNeeded),
+			s_format(""),
+			s_format("Trigger rate:"),
+			s_format("1 / %.3f", rateCapAdjusted),
+			s_format("= %.2f ^8per second", 1 / rateCapAdjusted),
+		}
+	end
+	return 1 / rateCapAdjusted
+end
+
+-- Calculate Trigger Rate impact due to other skills in rotation
+-- This is achieved by simulation a 10 second cast rotation
+local function calcMultiSpellRotationImpact(env, skillRotation, sourceAPS)
+	local SIM_TIME = 100.0
+	local index = 1
+	local time = 0
+	local next_trigger = 0
+	local trigger_increment = 1 / sourceAPS
+	local wasted = 0
+
+	while time < SIM_TIME do
+		local currIndex = index
+	
+		if time >= next_trigger then
+			while skillRotation[index].next_trig > time do
+				index = (index % #skillRotation) + 1
+				if index == currIndex then
+					wasted = wasted + 1
+					break
+				end
+			end
+
+			if skillRotation[index].next_trig <= time then
+				skillRotation[index].count = skillRotation[index].count + 1
+				skillRotation[index].next_trig = skillRotation[index].next_trig + skillRotation[index].cd
+				index = (index % #skillRotation) + 1
+				next_trigger = next_trigger + trigger_increment
+			end
+		end
+		-- increment time by server tick time passage
+		time = time + (1 / data.misc.ServerTickRate)
+	end
+
+	local mainRate = 0
+	local trigRateTable = { simTime = SIM_TIME, rates = {}, }
+	if wasted > 0 then
+		t_insert(trigRateTable.rates, { name = "Wasted_Need Higher_ICDR", rate = wasted / SIM_TIME })
```

Comment:
This seems pretty shoehorned in and I would love for a better way to represent it other than using columns for things they're not really designed for.  As a user I wouldn't expect to have to concatenate 3 columns to read the text.

Reply:
Re-did this to not be so ugly...